x-pack/plugins/fleet/server/services/actions/actions.ts

```diff
@@ -0,0 +1,293 @@
+): Promise<{
+  status: 'success' | 'failed' | 'mixed';
+  items: Array<{
+    status: 'success' | 'error';
+    // action_id
+    id: string;
+  }>;
+}> => {
+  const actions: FleetActionRequest[] = [];
+  const bulkCreateActionsBody = _actions.reduce<BulkCreate>((acc, action) => {
+    // doc id is same as action_id
+    const actionId = action.action_id || uuidV4();
+    acc.push({ create: { _index: AGENT_ACTIONS_INDEX, _id: actionId } });
+    const actionDoc = {
+      ...action,
+      action_id: actionId,
+      '@timestamp': action['@timestamp'] || new Date().toISOString(),
+    };
+    acc.push(actionDoc);
+    actions.push(actionDoc);
+
+    return acc;
+  }, []);
+
+  try {
+    const bulkCreateActionsResponse = (await esClient.bulk({
+      body: bulkCreateActionsBody,
+      refresh: 'wait_for',
+    })) as unknown as estypes.BulkResponse;
+
+    const responseItems = bulkCreateActionsResponse.items;
+
+    responseItems.forEach((item) => {
+      if (!item.create?.error) {
+        const id = item.create?._id ?? '';
+        const loggingInfo = getLoggingInfo({ id, actions });
+        auditLoggingService.writeCustomAuditLog({
+          message: `User created Fleet action [id=${id}, user_id=${loggingInfo.user_id}, input_type=${loggingInfo.input_type}]`,
+        });
+      }
+    });
+
+    const status = responseItems.every((item) => item.create?.error)
+      ? 'failed'
+      : responseItems.some((item) => item.create?.error)
+      ? 'mixed'
+      : 'success';
+
+    return {
+      status,
+      items: responseItems.map((item) => ({
+        status: item.create?.error ? 'error' : 'success',
+        id: item.create?._id ?? '',
+      })),
+    };
+  } catch (createBulkActionsError) {
+    throw new FleetActionsError(
+      `Error creating bulk actions: ${createBulkActionsError.message}`,
+      createBulkActionsError
+    );
+  }
+};
+
+export const getActionsByIds = async (
+  esClient: ElasticsearchClient,
+  actionIds: string[]
+): Promise<{
+  items: FleetActionRequest[];
+  total: number;
+}> => {
+  try {
+    const getActionsResponse = await esClient.search({
+      index: AGENT_ACTIONS_INDEX,
+      from: 0,
+      size: ES_SEARCH_LIMIT,
+      query: {
+        bool: {
+          filter: [
+            {
+              terms: {
+                action_id: actionIds,
+              },
+            },
+          ],
+        },
+      },
+    });
+
+    const actions = getActionsResponse.hits.hits.reduce<FleetActionRequest[]>((acc, hit) => {
+      if (hit._source) {
+        acc.push(hit._source as FleetActionRequest);
+      }
+      return acc;
+    }, []);
+
+    return {
+      items: actions,
+      total: actions.length,
+    };
+  } catch (getActionsByIdError) {
+    throw new FleetActionsError(
+      `Error getting action: ${getActionsByIdError.message}`,
+      getActionsByIdError
+    );
+  }
+};
+
+export const getActionsWithKuery = async (
+  esClient: ElasticsearchClient,
+  kuery: string
```

Comment:
I think we should move away from using kuery to more narrow parameters, see reasoning here: https://github.com/elastic/ingest-dev/issues/1753

Reply:
I think for simple `==` type of queries that will work @ashokaditya

Reply:
Yeah, that's a better approach with different methods on the client with specific search needs. I'll dig around for existing searches in the meantime.

Reply:
> Yeah, having a narrower set of search fields as parameter would also mean that we build the query in a limited way. In this case, I'm thinking we build a boolean query with a must filter list, for all given search field values. 

I like this and also checking the use cases that plugins need to query agents. If all we need is a simple list of == filters, that would be ideal.

Reply:
I had a chat about this with kibana core, and they suggested that it's fine to keep using `kuery`, but we should keep it as narrow as possible by adding validation on which fields can be used: https://github.com/elastic/ingest-dev/issues/1753#issuecomment-1588656888

Reply:
Yeah we can either validate the mapped fields or allow only fields that are being used today from plugins.